library/modules/Units.cpp

```diff
@@ -1437,6 +1437,73 @@ int8_t Units::getCasteProfessionColor(int race, int casteid, df::profession pid)
     return 3;
 }
 
+df::goal_type Units::getGoalType(df::unit *unit, int goalIndex)
+{
+    CHECK_NULL_POINTER(unit);
+
+    df::goal_type goal = df::goal_type::STAY_ALIVE;
+    if (unit->status.current_soul
+        && unit->status.current_soul->personality.dreams.size() > goalIndex)
+    {
+        goal = unit->status.current_soul->personality.dreams[goalIndex]->type;
+    }
+    return goal;
+}
+
+std::string Units::getGoalName(df::unit *unit, int goalIndex)
+{
+    CHECK_NULL_POINTER(unit);
+
+    df::goal_type goal = getGoalType(unit, goalIndex);
+    bool achieved_goal = isGoalAchieved(unit, goalIndex);
+
+    switch (goal)
+    {
+    default:
+    case df::goal_type::STAY_ALIVE:
+        return achieved_goal ? "Stayed Alive" : "Stay Alive";
```

Comment:
I'm wondering if maybe these strings should get added to the Structure XML files (though the "Start a Family" one might be tricky since it's apparently gender-specific).

Reply:
Not seeing any structure xml files. Can you point me at an example?

Reply:
Oh, you mean df.units.xml? goal_type is defined there, but I guess I need to add additional sub-elements to it?

Reply:
Yeah, enum items can have "attributes". [Here](https://github.com/DFHack/df-structures/blob/0.47.04-r1/df.buildings.xml#L21) is an example, and [here](https://github.com/DFHack/dfhack/blob/0.47.04-r1/plugins/remotefortressreader/building_reader.cpp#L81) is an example of code that reads it.
In this case, there could be two attributes for the unachieved and achieved names. Maybe `name` and `achieved_name` or `caption` and `achieved_caption`. The case for `START_A_FAMILY` could use "Parent", although I haven't checked what the game displays - you could replace "Parent" in this function dynamically, or use a placeholder in the attribute definition, which [some unit_thought entries](https://github.com/DFHack/df-structures/blob/0.47.04-r1/df.unit-thoughts.xml#L667) do.

This isn't required, of course, but it might make the function more future-proof - two new goals were added in 0.47.01, and if someone adds more to the XML files in the future, they'll be more likely to update attributes if they're right there.

Reply:
Another benefit of using enum attrs: I made the naming change Quietust suggested in https://github.com/DFHack/df-structures/pull/409#discussion_r467296259, and it didn't require any code changes on the DFHack side at all.